Add canRemain method to TargetPoolAllocationDecider to move shards from local to remote pool for hot to warm tiering

neetikasinghal opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
This pr adds canRemain method to TargetPoolAllocationDecider to move shards from local to remote pool for hot to warm tiering. Also, makes changes in the LocalShardsBalancer to not skip the shards tiered from hot to warm.

### Related Issues
Related PR: https://github.com/opensearch-project/OpenSearch/pull/14891/

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

neetikasinghal commented:
> ❌ Gradle check result for [e364d1e](https://github.com/opensearch-project/OpenSearch/commit/e364d1e7d5cef4d39f22314118490036e131cfba): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/43609/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

- [org.opensearch.gateway.RecoveryFromGatewayIT.testShardStoreFetchMultiNodeMultiIndexesUsingBatchAction](https://build.ci.opensearch.org/job/gradle-check/43609/testReport/junit/org.opensearch.gateway/RecoveryFromGatewayIT/testShardStoreFetchMultiNodeMultiIndexesUsingBatchAction/): flaky - https://github.com/opensearch-project/OpenSearch/issues/14304

- [org.opensearch.remotestore.WritableWarmIT.testWritableWarmBasic](https://build.ci.opensearch.org/job/gradle-check/43609/testReport/junit/org.opensearch.remotestore/WritableWarmIT/testWritableWarmBasic/): looks like this test doesnt have a search node: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/remotestore/WritableWarmIT.java#L97: this needs an update in the test.

- [org.opensearch.action.admin.cluster.allocation.ClusterAllocationExplainIT.testAllocationFilteringPreventsShardMove](https://build.ci.opensearch.org/job/gradle-check/43609/testReport/junit/org.opensearch.action.admin.cluster.allocation/ClusterAllocationExplainIT/testAllocationFilteringPreventsShardMove/) - i need to check this

jed326 commented on server/src/main/java/org/opensearch/action/admin/indices/tiering/TieringUtils.java:
```diff
@@ -0,0 +1,51 @@
+     * @return true if the index is a partial index, false otherwise
+     */
+    public static boolean isPartialIndex(final IndexMetadata indexMetadata) {
+        return FeatureFlags.isEnabled(FeatureFlags.TIERED_REMOTE_INDEX)
```
IMO the feature flag check should be in where this is called from instead of in the Util class.

neetikasinghal replied:
makes sense, will update this

neetikasinghal replied:
i tried this, however this is leading to complicated if else logic at multiple places, so keeping it as is for now.

jed326 replied:
It should just be an additional `&&` or `||` statement in the places that you're changing below anyways right? My concern here is that it makes the logic in `LocalShardsBalancer` not super clear without knowing that this feature flag exists in the util class. @harishbhakuni do you have a second opinion on this?

harishbhakuni replied:
+1 to this, I would also prefer using this feature flag logic directly in the `LocalShardsBalancer` for better clarity.

neetikasinghal replied:
yeah it made sense to me as well, but the logic was still not super readable. 
I have now moved the feature flag out of the function and added some comments to help make it more readable.

jed326 commented on server/src/test/java/org/opensearch/cluster/routing/allocation/TieringAllocationBaseTestCase.java:
```diff
@@ -0,0 +1,60 @@
+        System.setProperty(FeatureFlags.TIERED_REMOTE_INDEX, "true");
+    }
+
+    @AfterClass
+    public static void teardown() {
+        System.setProperty(FeatureFlags.TIERED_REMOTE_INDEX, "false");
+    }
```
Can we use `FeatureFlagSetter#set` here instead? And then cleanup should be handled by `OpenSearchTestCase`:
https://github.com/opensearch-project/OpenSearch/blob/1084ba9b167af8342be0acea075c83fa1673dcb1/test/framework/src/main/java/org/opensearch/test/OpenSearchTestCase.java#L242-L247

jed326 reviewed (commented):
Only some minor comments. Please add a changelog entry as well.

harishbhakuni commented on server/src/test/java/org/opensearch/cluster/routing/allocation/TieringAllocationBaseTestCase.java:
```diff
@@ -0,0 +1,47 @@
+import static org.opensearch.index.IndexModule.INDEX_TIERING_STATE;
+
+@SuppressForbidden(reason = "feature flag overrides")
+public abstract class TieringAllocationBaseTestCase extends RemoteShardsBalancerBaseTestCase {
```
Do we need this additional base class? why don't we just keep this single method in the same test class or in RemoteShardsBalancer base test class.

neetikasinghal replied:
yes i think it will be helpful as it can have methods specific for tiering.

harishbhakuni reviewed (approved):
Changes.. LGTM

jed326 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:
```diff
@@ -574,7 +576,11 @@ void moveShards() {
+            // - if feature flag is not enabled, skip, or
+            // - if feature flag is enabled and shard is not partial, skip
+            if (RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shardRouting, allocation))
+                && (!FeatureFlags.isEnabled(FeatureFlags.TIERED_REMOTE_INDEX) || (!isPartialShard(shardRouting, allocation)))) {
```
Small nit:
We could rewrite `!FeatureFlags.isEnabled(FeatureFlags.TIERED_REMOTE_INDEX) || (!isPartialShard(shardRouting, allocation))` as `!(FeatureFlags.isEnabled(FeatureFlags.TIERED_REMOTE_INDEX) && isPartialShard(shardRouting, allocation))`.

IMO that's clearer to read as code but maybe a little hard to describe as code comment so will leave that up to you.

neetikasinghal replied:
sure, will update this

jed326 commented:
Re-triggered gradle check

neetikasinghal commented:
I am planning to merge this in for 2.17~

@shwetathareja wanted to take your opinion on this and see if you have any concerns here!

@sohami would also mind taking a look at it?

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/indices/tiering/TieringUtils.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @opensearch.internal
+ */
+@ExperimentalApi
```
you dont need this, only needed for interfaces. opensearch.internal is sufficient.

neetikasinghal replied:
i see, updated

shwetathareja commented on server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:
```diff
@@ -574,7 +576,8 @@ void moveShards() {
             ShardRouting shardRouting = it.next();
 
-            if (RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shardRouting, allocation))) {
+            if (RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shardRouting, allocation))
+                && !(FeatureFlags.isEnabled(FeatureFlags.TIERED_REMOTE_INDEX) && isPartialShard(shardRouting, allocation))) {
```
[minor]: move this check to a private method in the class as it is repeated multiple times

shwetathareja replied:
Also adding this new && condition will impact for remote snapshots for which also REMOTE_CAPABLE will be set.

neetikasinghal replied:
> [minor]: move this check to a private method in the class as it is repeated multiple times

done

> Also adding this new && condition will impact for remote snapshots for which also REMOTE_CAPABLE will be set.

the new condition is && plus a ! , which means the current flow for remote snapshots will not be impacted unless feature flag is enabled for tiered remote index, and even if the feature flag is enabled, the condition will not be satisfied only in case when the shard has index setting with data locality set to PARTIAL as the partial shards are expected to go to the dedicated search nodes.

shwetathareja commented on server/src/test/java/org/opensearch/cluster/routing/allocation/ShardsTieringAllocationTests.java:
```diff
@@ -0,0 +1,108 @@
+import static org.opensearch.cluster.routing.RoutingPool.getIndexPool;
+import static org.opensearch.index.IndexModule.INDEX_STORE_LOCALITY_SETTING;
+
+public class ShardsTieringAllocationTests extends TieringAllocationBaseTestCase {
```
Add test for searchable Snapshots as well

neetikasinghal replied:
added test `testShardsInRemotePool`

shwetathareja reviewed (commented):
Thanks @neetikasinghal for the changes. Left some comments.

neetikasinghal commented:
> Thanks @neetikasinghal for the changes. Left some comments.

@shwetathareja thanks for your review, I have addressed your comments, please take a look

shwetathareja commented on CHANGELOG.md:
```diff
@@ -40,13 +40,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Add index creation using the context field ([#15290](https://github.com/opensearch-project/OpenSearch/pull/15290))
 - Add fieldType to AbstractQueryBuilder and FieldSortBuilder ([#15328](https://github.com/opensearch-project/OpenSearch/pull/15328)))
 - [Reader Writer Separation] Add searchOnly replica routing configuration ([#15410](https://github.com/opensearch-project/OpenSearch/pull/15410))
+<<<<<<< HEAD
```
please clean up this

neetikasinghal replied:
updated

neetikasinghal commented:
> Thanks @neetikasinghal for the changes. Left some comments.

@shwetathareja thanks for your review, addressed your comments.

neetikasinghal commented:
this change is also present in 2.17
